Add auto mapper code and UI

code-monet opened the pull request:
I leaned on AI for the UI and it's not great, but might be good enough to do the job. Open to suggestions now, or we can come back and improve it later.

Fixes #238

code-monet committed:
Add a NamedTuple to represent a distinct vJoy input

code-monet committed:
Add some type hints to profile.py

code-monet committed:
Add auto mapper library, unit tests

code-monet committed:
Add dialog for the auto mapper

WhiteMagic commented on gremlin/profile.py:
```diff
@@ -722,12 +722,12 @@ def __init__(self, library: Library) -> None:
-        self.mode = None
+        self.device_id: uuid.UUID | None = None
+        self.input_type: InputType | None = None
+        self.input_id: int | tuple[int, int] | None = None
```
Int is for joystick devices and the tuple for keyboard keys correct?

code-monet replied:
Correct!

WhiteMagic commented on gremlin/ui/backend.py:
```diff
@@ -312,6 +313,10 @@ def getInputItem(
     def getLogicalDeviceManagementModel(self) -> LogicalDeviceManagementModel:
         return LogicalDeviceManagementModel(self)
+    
+    @Slot(result=AutoMapper)
```
Did you experiment with just creating the object directly in QML rather than going via the backend instance? That should likely work as usually the backend detour is only needed for objects that require some odd params during construction.

code-monet replied:
Changed.

I asked AI several times and it kept saying I had to go via `engine.rootContext().setContextProperty`... but now I see how you've been using `QtQml.QmlElement` and did the same thing.

WhiteMagic replied:
The `QtQML.QmlElement` decorator is a shortcut to exporting a QML object to the engine. It does what the `PluginManager` does
```
QtQml.qmlRegisterType(plugin.create.model, "Gremlin.ActionPlugins", 1, 0, plugin.create.model.__name__)
```
with the decorator. Though I may at some point replace all of them with the specific format as linters don't understand what goes on with that decorator and then give you all kinds of false positive complaints.

code-monet replied:
Ack (no action needed, right?)

WhiteMagic replied:
Yep all good, I just added the explanation in case you were wondering what that thing does.

WhiteMagic commented on gremlin/ui/device.py:
```diff
@@ -259,6 +260,7 @@ class DeviceListModel(QtCore.QAbstractListModel):
         "vid": lambda dev: "{:04X}".format(dev.vendor_id),
         "guid": lambda dev: str(dev.device_guid),
         "joy_id": lambda dev: dev.joystick_id,
+        "vjoy_id": lambda dev: dev.vjoy_id,
```
Is this safe for non vJoy devices?

code-monet replied:
It's "safe", yes, you'll just get the placeholder -1 for those.

WhiteMagic replied:
Ok as long as it doesn't horribly crash, though I also don't think that value should be queried by anything that doesn't already know it deals with vJoy devices.

WhiteMagic commented on qml/Main.qml:
```diff
@@ -288,6 +292,23 @@ ApplicationWindow {
             }
 
+            ToolButton {
+                text: "1:1"
```
If this actually just renders the literal text "1:1" the whole bootstrap icon part isn't needed. That is only of use if you want to use an icon that's embedded in that font.

code-monet replied:
Remove that section.

WhiteMagic commented on vjoy/vjoy.py:
```diff
@@ -68,6 +68,14 @@ class HatType(enum.Enum):
     Continuous = 1
 
 
+class VjoyInput(NamedTuple):
```
As the vjoy module isn't using this on its own this isn't the most ideal place. Since it is only used with the auto_mapper feature I'd probably place it in that module. Technically we could reuse the `InputIdentifier` but that is tied to Qt so not actually useful for non-ui code.

If the thought is that there might be a need for a generic "deviceidentifier" class it would also be an option to put this into types and possible generalize to general joystick devices by using the `DeviceType` enum to encode the type of input (and extend it to include Logical as well as vjoy, joystick, and keyboard)

code-monet replied:
I was looking at `map_to_vjoy` and thought that this might be useful there. I wouldn't extend it to all devices since the device identifier type then gets overloaded, and same for input identifier (e.g. with keyboard key code tuples). Going by that, what's your preference between a) keeping it in vjoy and in the future re-use in `map_to_vjoy` (and `macro` perhaps), or b) move to `auto_mapper` (or c) move to `types` if you wish) ?

WhiteMagic replied:
Keeping it to vJoy is fine since it's a bit of an odd thing. I would not put it in `auto_mapper` as that's not something that shouldn't need to be imported just to get this type. I think `types` might be the cleanest place for this. My main reason for not wanting it inside of `vjoy` is that I'd like to keep that module as Gremlin free as possible. Such that if in the future it is replaced or another direct input emulation driver comes along there aren't many threads to untangle.

Since this "type" is just a named tuple it won't introduce crazy import dependencies which could cause circular imports, which also makes it a safe candidate for types.

code-monet replied:
Sounds reasonable, moved to `types`.

WhiteMagic commented:
I'll take a look at the mapper logic and UI stuff later. One thing I saw is that the UI here uses mostly row/column while I mostly tend to use rowlayout and collayout. I still don't like either as both kind of are dumb in different ways :-) For a simple UI both should be fine, I'll possibly even pull the whole thing into a separate branch to be able to actually run it as with UI code staring at the XML only tells you so much with QML.

WhiteMagic commented:
On the QML front, while it works, there are quite a few things I personally would change.

- Place options checkboxes below the device selections and replace them with Slider widgets (I'm not the biggest fan of them but that's where Windows is going)
- I try to use `_camelCase` for identifiers and only use them for components I need to refer to
- Use `RowLayout` and `ColumnLayout` instead of `Row` and `Column` as the anchor stuff gets quite messy and just contains magic values with no end
  - You can `anchors.fill: parent` in the main `ColumnLayout` and then have individual rows contain more `ColumnLayout` with `Layout.fillWidth: true` in both instances to get a 50/50 split.
-  The LLM also seemed to really like overly long identifier names
- Repeaters work but with a proper `AbstractListModel` it is better (usually) to use a `ListView` qml widget (there are instances where this doesn't work but I think this shouldn't be one of them)
- I don't have oversized buttons anywhere so there shouldn't be any here either
- Gremlin uses the Universal style (https://doc.qt.io/qt-6/qtquickcontrols-universal.html) and sticks to the colors it defines, one value that the page doesn't list is `Universa.baseLowColor` that is a useful "background" color. So I wouldn't define random hex values. For the `TextArea` I'd see if a simple line box is also suitable (such that it works on the same line as the non-oversized button) and then could also be turned into a separate "notification" widget so the style can be consistent and reused.
- I think having a short "description/manual" at the top explaining what/how to use the tool is fine. Alternatively you could add a `?` button on the bottom row with the button and status display.

code-monet committed:
First pass improvements to the automapper GUI

code-monet commented:
Apologies that I intended to and forgot to attach a video of the UI - will do so for the next push so you don't have to spin up the UI each time. I'm fixing up the UI now; previously I did try to use the `Layout` variants of Row and Column, but the device checkboxes weren't clickable and didn't show any text, and I wasn't able to fix it. I'll try it again, referring to your other QMLs.

---

(Converting your list to checkboxes for my own tracking)

- [x]     Place options checkboxes below the device selections and replace them with Slider widgets (I'm not the biggest fan of them but that's where Windows is going)
- [x]     I try to use _camelCase for identifiers and only use them for components I need to refer to
- [x]     Use RowLayout and ColumnLayout instead of Row and Column as the anchor stuff gets quite messy and just contains magic values with no end
- [x]         You can anchors.fill: parent in the main ColumnLayout and then have individual rows contain more ColumnLayout with Layout.fillWidth: true in both instances to get a 50/50 split.
- [x]     The LLM also seemed to really like overly long identifier names
- [x]     Repeaters work but with a proper AbstractListModel it is better (usually) to use a ListView qml widget (there are instances where this doesn't work but I think this shouldn't be one of them)
- [x]     I don't have oversized buttons anywhere so there shouldn't be any here either
- [x]     Gremlin uses the Universal style (https://doc.qt.io/qt-6/qtquickcontrols-universal.html) and sticks to the colors it defines, one value that the page doesn't list is Universa.baseLowColor that is a useful "background" color. So I wouldn't define random hex values. For the TextArea I'd see if a simple line box is also suitable (such that it works on the same line as the non-oversized button) and then could also be turned into a separate "notification" widget so the style can be consistent and reused.
- [x]     I think having a short "description/manual" at the top explaining what/how to use the tool is fine. Alternatively you could add a ? button on the bottom row with the button and status display.

WhiteMagic commented:
No worries about the video it's easy enough to locally checkout the PR and run it.

For things not showing up when using layouts, there is a lot of weird quirky stuff going on with them at times. I often use the `DebugBox { target: _idOfSomething` } to visualize the space something takes up. With layouts it's important to give them size as normal with anchors but then their direct content needs to use `Layout.something` for dimensions as otherwise QML complains. The layouts also figure out how "big" something is based on the `implicitWidth` and `implicitHeight` of items. So sometimes it is necessary to compute those values and you can check if that's the issue by forcibly setting that value to something.

Also if you get stumped by it just commit what you have and let me know, I dislike UI stuff a lot but I've been dealing with a lot such that I know quite a few of the typical non-obvious ways to shoot yourself into the foot with them. Als LLMs seem to get better at knowing some of QML's quirks but sometimes they are also just totally wrong with their outputs.

code-monet commented:
Okay, I'm taking a short break but this is what I have so far (see checkboxes above). Additional notes:

1. I think you meant Switch and not Slider widget?
2. I didn't use baseLowColor for the notification box since it was matching the button, and the button color currently matches what I see e.g. in Actions

WhiteMagic commented:
Wow that's much nicer :-)

And yes I meant switch, I always get confused with what their name is in QML. Color use is fine, I just mentioned it in case it's useful because the documentation doesn't really say that the variable exists.

For the `ListView` you don't need to place them inside a `ScrollView` as that's doubling things up. `ListView`  is a scrollable container on its own. It behaves a bit dumb, so I'll actually make a `JGListView` or similar so I don't have to repeat the same things in them repeatedly.  The usual add-ons I do are:

```
ListView {
  clip: true // otherwise you can see the content outside the widget's boundariers

  // Always show a scrollbar, depends on use case
  ScrollBar.vertical: ScrollBar {
    policy: ScrollBar.AlwaysOn
  }
  // No bouncing or drag to move, i.e none of the phone behavior stuff
  flickableDirection: Flickable.VerticalFlick
  boundsBehavior: Flickable.StopAtBounds
}
```

For the `Switch` widgets you can also use the `checked` property of theirs in the `onToggled` which is directly the boolen you want to assign, i.e. no need to deal with the `position` and odd "truthness" things.

You don't need the `Rectangle` as the top element to get the spacing around the content. You can just add `anchors.margins: 10` to the to the top-level `ColumnLayout` to get the same effect.

The two `Switch` items could also be in a `RowLayout` of their own and split the space as they see fit, as opposed to being aligned with one of the two boxes, but that's really just another option and I don't have opinions on either layout.

Not all `Layout.fillX` are needed, but I honestly don't know the exact rules around what QML assumes as defaults or  the logic employed. There is this snippet in the documentation but it's kind of vague on what exactly will happen, especially when no `Layout.fillX` or `Layout.preferredX` is given.

> The [fillWidth](https://doc.qt.io/qt-6/qml-qtquick-layouts-layout.html#fillWidth-attached-prop) and [fillHeight](https://doc.qt.io/qt-6/qml-qtquick-layouts-layout.html#fillHeight-attached-prop) properties can either be true or false. If they are false, the item's size will be fixed to its preferred size. Otherwise, it will grow or shrink between its minimum and maximum size as the layout is resized. If there are multiple items with [fillWidth](https://doc.qt.io/qt-6/qml-qtquick-layouts-layout.html#fillWidth-attached-prop) (or [fillHeight](https://doc.qt.io/qt-6/qml-qtquick-layouts-layout.html#fillHeight-attached-prop)) set to true, the layout will grow or shrink the items relative to the ratio of their preferred size.

If you want I can also quickly implement the changes I mention into the QML file and probably just paste it here as making a PR against a PR might be funny, never tried that before.

**Edit:** Created that new `ListView` version `JGListView` which has the correct behavior by default and has a property `scrollbarAlwaysVisible` which allows toggling that on/off for the vertical scrollbar.

code-monet committed:
Remove redundant items from DialogAutomapper.qml and use the new JGListView

code-monet commented:
Done:

- [x] Using JGListView (and rebased on top of latest dev). Thanks for adding that!
- [x] Using `checked` property for `Switch`es
- [x] Removed top level Rectangle (AI apparently likes using more code than necessary)
- [ ] The Switches I left aligned with their respective device lists - that alignment/relationship makes sense to me
- [x] Removed some `FillWidth` and `FillHeight`s

---

I'm fine any way for QML/code suggestions - as a post in the PR, pasted snippet, separate branch that I can peek at - but really your posts have enough detail that I have been able to implement the changes without too much trouble :)

WhiteMagic commented on gremlin/profile.py:
```diff
@@ -507,12 +507,12 @@ class Profile:
         self.modes = ModeHierarchy(self)
         self.scripts = ScriptManager(self)
-        self.fpath = None
+        self.fpath: str | None = None
```
I'm really tempted to make a wrapper around Path only to be able to have an "invalid" path :-)

code-monet replied:
In theory the `exists()` function is what you'd use to determine if it's "valid", whereas there's still a separate concept, of whether one has been specified (yet), for which using `| None` is fine, IMO.

WhiteMagic replied:
The trouble is with an empty path becoming just `.` rather than being invalid as is the case in C++ :-( But yeah it's just a general annoyance but not too common to fix ... for now at least.

code-monet replied:
Definitely re: empty strings. Marking as resolved, something to consider in the future.

WhiteMagic commented on gremlin/ui/auto_mapper.py:
```diff
@@ -0,0 +1,70 @@
+from gremlin import auto_mapper
+from gremlin.ui import backend
+
+QML_IMPORT_NAME = "Gremlin.UI"
```
I would suggest making a new package here, maybe `Gremlin.Tools` such that we have a nice group for everything that's tool-like. In the future the device remapper will need to live somewhere too and that'd be a good label I think.

code-monet replied:
Good idea, done.

WhiteMagic commented on gremlin/ui/auto_mapper.py:
```diff
@@ -0,0 +1,70 @@
+        super().__init__(parent)
+
+    @QtCore.Slot(dict, dict, bool, bool, result=str)
+    def create_mappings(self, physical_devices, vjoy_devices, overwrite, repeat) -> str:
```
Looking at this, since this is just a single function we really use, might also just put that into the `Backend` class. The Backend is getting bloated over time but that's kind of its purpose I guess.

If you needed more actual interaction with the UI and expose model data it would make sense to have a dedicated model, but since all you want to do is a "one shot" function execution without storing or exposing state I think the backend route is more "normal" for Gremlin. Though "normal" is changing over time too :-)

code-monet replied:
That would be less code, but I agree with your concern for bloat in the Backend, which is not just a "backend" but also serving the purpose of a namespace. I'd keep this as a separate file just to keep things cleaner and reduce how many things are imported into `backend.py`.

And who knows, we might end up making the auto_mapper implementation/UI more complicated/feature-full in the future!

WhiteMagic commented on gremlin/ui/auto_mapper.py:
```diff
@@ -0,0 +1,70 @@
+        Returns:
+            A string report for the user summarizing new mappings.
+        """
+        logging.getLogger("system").info(
```
I tend to use f"{variable}" strings for formatting everywhere as opposed to the really old "%s" substitution style.

code-monet replied:
There is a performance benefit to doing it the old way in logging calls - the string isn't formatted unless the log line actually needs to be generated.

However since Gremlin doesn't really change log levels this aspect won't come into play - so I have changed it to f-strings.

WhiteMagic replied:
I don't tend to care for performance aspects unless it's a very obvious case where it will matter, so usually tight loops of compute or reacting to user inputs in even handling. This being a tool, I think there can be a lot of "slow" code and no one would notice.

WhiteMagic commented on qml/TextOutputBox.qml:
```diff
@@ -0,0 +1,36 @@
+Rectangle {
+    id: root
+    property alias text: _notificationLabel.text
+    
```
Possibly good to add 
```
implicitHeight: _notificationLabel.height
implicitWidth: _notificationLabel.width
```
What this does is make the Rectangle have the size large enough to show the content of the label. Without that it won't have a size, i.e. nothing shown, unless the user specifies one. Using the `implicit` values allows Layouts to use that information as a hint for the required size when they do their layouting.

So what should happen in your code if you remove the `Layout.fillWidth` etc. is that the text should still show up without overlapping other things because it knows how big it is. Without the additional information you actually wouldn't see the text. At least that's what I think should happen, which with QML oftentimes isn't super clear.

code-monet replied:
I have added these, but it didn't have the desired effect: `layout.fillWidth` is still needed. Per AI, this is expected:

"The issue is that the TextOutputBox relies on its implicit dimensions, but in a RowLayout, the layout needs explicit constraints to properly size the component. 

Without an explicit Layout.preferredHeight, the RowLayout collapses the height to 0 because:
* The TextOutputBox's implicit height depends on its content
* But the content's height depends on the component's height (circular dependency)
* The layout can't resolve this, so it collapses to 0 height"

WhiteMagic replied:
I'd say the LLM is inventing things there, the documentation about layouts has this to say.

> Note: Do not bind to the x, y, width, or height properties of items in a layout, as this would conflict with the goals of Layout, and can also cause binding loops. The width and height properties are used by the layout engine to store the current size of items as calculated from the minimum/preferred/maximum attached properties, and can be ovewritten each time the items are laid out. Use [Layout.preferredWidth](https://doc.qt.io/qt-6/qml-qtquick-layouts-layout.html#preferredWidth-attached-prop) and [Layout.preferredHeight](https://doc.qt.io/qt-6/qml-qtquick-layouts-layout.html#preferredHeight-attached-prop), or [implicitWidth](https://doc.qt.io/qt-6/qml-qtquick-item.html#implicitWidth-prop) and [implicitHeight](https://doc.qt.io/qt-6/qml-qtquick-item.html#implicitHeight-prop) to specify the preferred size of items.

However, once you start nesting things the lines get very blurry and stuff might just misbehave and I have mostly given up trying to fully understand the rhyme and reason behind the behavior. If it doesn't work we can omit it or leave it in in the hopes that maybe it works in other situations. Though realistically in most cases one will specify dimensions anyhow.

So no need to waste time trying to make it behave.

code-monet replied:
I'll leave it in, perhaps one day in a different QML we'll find that it works and then we'll have solved the mystery!

WhiteMagic commented:
> Done:
> 
>     * [x]  Using JGListView (and rebased on top of latest dev). Thanks for adding that!
> 
>     * [x]  Using `checked` property for `Switch`es
> 
>     * [x]  Removed top level Rectangle (AI apparently likes using more code than necessary)
> 
>     * [ ]  The Switches I left aligned with their respective device lists - that alignment/relationship makes sense to me
> 
>     * [x]  Removed some `FillWidth` and `FillHeight`s
> 
> 
> I'm fine any way for QML/code suggestions - as a post in the PR, pasted snippet, separate branch that I can peek at - but really your posts have enough detail that I have been able to implement the changes without too much trouble :)

Nice that looks good to me. I'll try to go over the main auto mapper code in the evening, didn't manage to look at that one yet.

WhiteMagic commented on gremlin/auto_mapper.py:
```diff
@@ -0,0 +1,296 @@
+        self, input_devices: list[dill.DeviceSummary], options: AutoMapperOptions
+    ) -> Iterable[profile.InputItem]:
+        """Iterates over physical axes that need to be mapped in a prepared profile."""
+        current_mode = mode_manager.ModeManager().current.name
```
This won't give you the currently selected mode in the UI. The mode manager only serves a function when Gremlin is running. The UI selection information is only held in the `UIState` object but that one exists only as an instance within the UI and isn't exposed in any way.

In short, there is no way at the moment for code to know what the currently selected mode is without the UI passing that information on. If that information is needed I see two ways:
1. Devise a mechanism to expose UI state to Gremlin code as well.
2. Have the UI pass the name of the active mode on to code it calls.

Realistically most code shouldn't need to be able to query the UI state, though here I could see that being necessary. Though exposing it could become quite ugly so I probably lean more towards letting the UI forward that information.

code-monet replied:
I will follow your suggestion of adding mode to a drop down in the UI.

WhiteMagic commented on gremlin/auto_mapper.py:
```diff
@@ -0,0 +1,296 @@
+    def _get_used_vjoy_inputs(self) -> list[vjoy.VjoyInput]:
+        """Returns a list of all vJoy inputs that are already used in the prepared profile."""
+        used_vjoy_inputs = []
+        connected_device_uuids = set(
```
Did you run across instances where a device is listed multiple times in that list? If so that's a bug as it makes no sense for the same device to exist more than once. So `set` should be superfluous, if it's needed the other code has bugs.

code-monet replied:
Not needed, this is a leftover from earlier logic - removed set cast.

WhiteMagic commented on gremlin/auto_mapper.py:
```diff
@@ -0,0 +1,296 @@
+        )
+        for device_uuid, input_items in self._profile.inputs.items():
+            if device_uuid not in connected_device_uuids:
+                logging.getLogger("system").debug(
```
This may become very spammy and I'd probably just skip this.

code-monet replied:
Removed.

WhiteMagic commented on gremlin/auto_mapper.py:
```diff
@@ -0,0 +1,296 @@
+                )
+                continue
+            for input_item in input_items:
+                if input_item.mode != mode_manager.ModeManager().current.name:
```
As mentioned above, this is not the same as the mode selected in the UI. Another option would be to have a mode selector dropdown in the 1:1 mapping dialog. That way it's actually clear to which mode the mappings are being added, as opposed to the user needing to figure out that it's the currently selected one.

code-monet replied:
Good idea, added drop down in the tool UI.

WhiteMagic commented on gremlin/auto_mapper.py:
```diff
@@ -0,0 +1,296 @@
+                    )
+                    continue
+                for binding in input_item.action_sequences:
+                    assert isinstance(binding.root_action, root.RootData)
```
Unless someone manually messed with the XML you can't actually have the root_action be anything other than Root. Though there is nothing saying there couldn't be other options in the future if there is a need.

code-monet replied:
I just like having such a lightweight internal check wherever there's an assumption being made about a type (that the type checker can't determine). It makes error messages easier to understand (mainly intended for developers).

WhiteMagic replied:
I generally agree, though here it is currently an invariant that must hold, but in the future this may not be true. I guess it might be fine to run into this issue at that time, though.

WhiteMagic commented on gremlin/auto_mapper.py:
```diff
@@ -0,0 +1,296 @@
+                for binding in input_item.action_sequences:
+                    assert isinstance(binding.root_action, root.RootData)
+                    child_actions = binding.root_action.children
+                    for child_action in child_actions:
```
Could be simplified to: `for child_action in binding.root_action.children:`.

code-monet replied:
Done.

WhiteMagic commented on gremlin/auto_mapper.py:
```diff
@@ -0,0 +1,296 @@
+                        else:
+                            logging.getLogger("system").debug(
+                                f"Auto-mapper will reuse vJoy mappings in {child_action=}"
+                            )
```
Not sure what the log is meant to mean, and similar to the earlier one would just result in a lot of spam in the log file.

I think there is also a logic error here. The code assumes that all vJoy maps are children of the root node, that is not true. In R14 you can have arbitrarily deep nestings and each action may have "children", e.g. the condition action has children actions it runs when the condition holds and another set when they don't. Using `children` on the root action is also using knowledge/information that may change.

The `base_classes.AbstractActionData` class, from which every action has to inherit, does provide the basic building block you'd have to use to get all child actions of a given action via the `get_actions(selector: str|None=None)`. If no argument is given it will return all children of that action, irrespective of the "child container" they live in. That interface must be implemented by every action that can have children, otherwise the UI breaks, so you using that will always give you the right actions. Now what this does not do is recursively go through the children you get in that way and get their children, if they  have any. So if you have the root action the following should give you all actions that are in the possibly very deep nesting of actions:

```
all_actions = []
candidate_actions = root_action.get_actions()
while len(candidate_actions) > 0:
    all_actions.append(candidate_actions.pop(0))
    candidate_actions.extend(all_action[-1].get_children())
```
This should be correct but testing wouldn't hurt :-)

Once you have the list of all actions you can filter them or you could already filter them in there if you wanted. For the filtering there is another approach you could use. The `Library` has a method `actions_by_type` which returns all action data instances that match a given type. You could then use that to do a `action in vjoy_actions` check, though tbh. I have no idea which one is faster or more convenient.

code-monet replied:
Replaced log with comment.

I realized my class document didn't match implementation, updated that; based off our earlier conversation in #238 I decided to keep things simple and only consider a vJoy input "used" if it was in a direct mapping. The idea being, a user could run the tool, remove/change some of the generated bindings, and re-run the tool.

For more complicated logic, I propose punting them to the future until someone asks for it (and probably adding a switch in the UI, like we have for overwriting existing physical inputs).

WhiteMagic commented on gremlin/auto_mapper.py:
```diff
@@ -0,0 +1,296 @@
+    ) -> Iterable[vjoy.VjoyInput]:
+        """Returns a list of all vJoy inputs that are not used in the prepared profile."""
+        for vjoy_dev in device_initialization.vjoy_devices():
+            if vjoy_dev.vjoy_id not in vjoy_ids:
```
Do you know what happens with this if someone has vjoy device 1 and 3 but not 2? It's an odd edge case but some people do this for reasons.

code-monet replied:
It should be fine, at least the UI is correctly showing and passing the VJoy device ID in such a configuration, and none of the code here makes an assumption about continuity (at least going by `range` calls...)

WhiteMagic replied:
Ok then that should be fine :+1:

WhiteMagic commented on gremlin/auto_mapper.py:
```diff
@@ -0,0 +1,296 @@
+                    vjoy_dev.vjoy_id, types.InputType.JoystickAxis, axis_index
+                )
+                if vjoy_axis not in used_vjoy_inputs:
+                    yield vjoy_axis
```
How does this behave when you can't yield because you "ran out of" unused inputs?

code-monet replied:
No more bindings will be created for that input type, unless "repeat VJoy devices" is checked, in which case the inputs will be looped as discussed in #238 

Tangentially, I realize that a lot of edge cases are possible and that the unit tests don't really cover many of those... I can come back and add more unit tests in a future PR, using mock devices rather than a profile (which for the time being is nice, because I can load it up and view it in Gremlin).

WhiteMagic replied:
Sounds good to me, no need to overdo the bullet proofing :-)

code-monet replied:
Thanks, marking as resolved.

WhiteMagic commented on gremlin/auto_mapper.py:
```diff
@@ -0,0 +1,296 @@
+        binding.root_action = root_action
+        binding.behavior = physical_input.input_type
+        physical_input.action_sequences.append(binding)
+        self._created_mappings.append(vjoy_action)
```
As this is mostly a copy of what `gremlin.ui.profile.InputItemModel.newActionSequence` does I think it might make sense to move the logic of creating a new `InputBinding` into the actual `gremlin.profile.InputItem` class. That way we remove the duplication and more importantly if there is ever anything wrong with it or changed both uses are automatically fixed.

Moving this should be straight forward I think. In the `InputItemModel` code the call to the new method just needs to be in between the `beginInsertRows()` and `endInsertRows()` calls.

code-monet replied:
Good call, added `add_item_binding`

WhiteMagic commented on gremlin/auto_mapper.py:
```diff
@@ -0,0 +1,296 @@
+    profile,
+    shared_state,
+    types,
+)
```
I've mostly just done list without the tuple notation and broken it with `\` when it exceeded 80 chars in length.

code-monet replied:
Changed, although the `black` formatter prefers this. What formatter are you using? We could specify it in `contributingmd` and configure it via the TOML, e.g. https://developer.lsst.io/python/formatting.html#id1

WhiteMagic commented on gremlin/auto_mapper.py:
```diff
@@ -0,0 +1,296 @@
+from vjoy import vjoy
+
+
+@dataclasses.dataclass
+class AutoMapperOptions:
+    """Options for the auto-mapper.
+
+    More flags may be added here to further customize auto-mapper behavior.
+    """
+
+    repeat_vjoy_inputs: bool = False
+    overwrite_used_inputs: bool = False
```
To me it feels a bit overkill to use a class for just two params. However, if we'd need to pass the mode name around as well then I think it starts to be sensible to pack them into one single thing.

code-monet replied:
Indeed I did it with a class because I expect we'll end up adding more things to it.

WhiteMagic commented on gremlin/auto_mapper.py:
```diff
@@ -0,0 +1,296 @@
+    to it in the profile, but:
+    1. The binding is from a disconnected device.
+    2. The binding is from a macro, a sub-action (usually temp, chain, or condition), or in a
+       user script.
```
Hmm this case I probably wouldn't exclude. If someone has those they may not accidentally want them to be used. But only if its part of an active device obviously.

code-monet replied:
I had coded that up partially earlier, but then removed it because it was getting complicated (also needed knowledge of how macro, chain etc using vJoy). Per our [earlier thread](https://github.com/WhiteMagic/JoystickGremlin/issues/238#issuecomment-3337092915) perhaps we can come back and modify this later if needed? I agree with our earlier discussion that it's unlikely someone will add somewhat advanced bindings and *then* run the tool.

WhiteMagic commented on gremlin/auto_mapper.py:
```diff
@@ -0,0 +1,296 @@
+
+    A physical input is considered "available" if it has no binding in the profile. An option
+    is provided to overwrite unavailable inputs, in which case any existing bindings are removed,
+    unless they are direct vJoy mappings (which could have been created by this tool).
```
Just a thought occurred, you need to check that the root action has 0 children, if someone clicked on an item sometimes Gremlin initializes things but they aren't "there", so assuming anything that has an `InputItemBinding` is used may not be true.

code-monet replied:
I had it coded up like that originally, but simplified it because it didn't seem like a thing that could happen. Perhaps instead of accommodating that here, would it be better to track down situations where empty InputItemBindings are created and prevent that from happening?

WhiteMagic commented on gremlin/auto_mapper.py:
```diff
@@ -0,0 +1,296 @@
+
+    Bindings are generated for the current active mode only; bindings from other modes are
+    not checked (i.e. any physical and vJoy inputs used only in other modes are considered
+    available in the current mode).
```
As mentioned below, the code can't do this at the moment. I ran into the same problem when working on the condition action again. Only there I actually don't care and could ignore this issue.

code-monet replied:
I'll put mode into a drop down in the tool UI.

WhiteMagic commented on gremlin/ui/auto_mapper.py:
```diff
@@ -0,0 +1,70 @@
+        super().__init__(parent)
+
+    @QtCore.Slot(dict, dict, bool, bool, result=str)
+    def create_mappings(self, physical_devices, vjoy_devices, overwrite, repeat) -> str:
```
This should be `createMappings` the reason is that Qt uses `camelCase` for everything and while it's not the usual Python naming convention I use it when PySide6 code exposes something as a `Slot` or `Property`. That way it's directly clear that the method deals with QML code and on the QML side it doesn't look out of place.

code-monet replied:
Renamed.

WhiteMagic commented on gremlin/auto_mapper.py:
```diff
@@ -0,0 +1,296 @@
+        Returns:
+            A string report for the user summarizing new mappings.
+        """
+        input_devices = [
+            dev
+            for dev in device_initialization.physical_devices()
+            if dev.device_guid in input_devices_guids
+        ]
```
Ah the `DeviceSummary` objects are using `dill.GUID` instead of `uuid.UUID` correct? I wouldn't change that only asking so no awkward bugs get in.

code-monet replied:
Yes, which is probably fine - but in a future PR I can switch `device_initialization` to use UUID instead.

WhiteMagic replied:
Yes for now this is totally fine. In the future it's probably easiest/safest to not even expose the GUID via the `DeviceSummary`. That way it can't "escape" the `dill` module.

code-monet replied:
If we implement the ability to copy bindings from one device to another, then I can't think of any other good reason to make the GUID available to the user. (marking as resolved).

WhiteMagic commented on gremlin/ui/auto_mapper.py:
```diff
@@ -0,0 +1,70 @@
+            [int(vjoy_id) for (vjoy_id, chosen) in vjoy_devices.items() if chosen],
+            auto_mapper.AutoMapperOptions(
+                repeat_vjoy_inputs=repeat,
+                overwrite_used_inputs=overwrite,
```
Do you have to use named parameters or does it also allow you to just use them in the positional order?

code-monet replied:
Positional will also work, switched to that.

WhiteMagic commented:
Ok had a look over the mapping code, overall looks correct and good to me, despite the lengthy comments :-)

code-monet committed:
Second pass improvements to auto mapper code and GUI

code-monet commented:
Okay, I *think* I have responded to all comments - let me know what you think, I can make more changes depending on what we decided we want to put into this first implementation.

I re-used the TextOutputBox but something is still missing because it needs either `fillWidth` or `preferredWidth`, otherwise it collapses to 0.

WhiteMagic commented:
I had one other though on the ui mode bit. It could be possible to store the UI mode also in the `ModeManager` as it is Qt enabled. However, the drawback/ugliness of this is that Qt code gets dragged into otherwise Qt free code. I try to keep everything outside the `gremlin/ui` folder free of Qt stuff.

WhiteMagic commented on gremlin/auto_mapper.py:
```diff
@@ -0,0 +1,276 @@
+
+
+@dataclasses.dataclass
+class AutoMapperOptions:
```
Since we have this data class I'd put the `mode` value into this as well, otherwise, it's a bit odd to pass a param struct but then still need to pass around more params outside of it.

code-monet replied:
Moved to options.

code-monet committed:
Move mode to options in automapper, and fix a bug where mode wasn't being correctly passed in if the user never selected one

code-monet commented:
>  It could be possible to store the UI mode also in the ModeManager as it is Qt enabled. However, the drawback/ugliness of this is that Qt code gets dragged into otherwise Qt free code. I try to keep everything outside the gremlin/ui folder free of Qt stuff.

I don't use the ModeManager anymore, the dropdown idea is good! More generally, it seems reasonable to keep UI and non-UI stuff separated.

WhiteMagic commented:
Very nice, good to merge from my side. Let me know if you wanted to add anything else in, otherwise I'll merge it.

code-monet commented:
Go ahead with the merge, I'll send a new PR with some minor improvements.
